packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go

```diff
@@ -669,23 +801,116 @@ func (dhtPeer *DHTPeer) RouteEnvelope(envel *aea.Envelope) error {
 
 		linfo().Str("op", "route").Str("addr", target).
 			Msgf("sending envelope to target peer %s...", peerID.Pretty())
-		err = utils.WriteEnvelope(envel, stream)
+
+		envelBytes, err := proto.Marshal(envel)
 		if err != nil {
+			lerror(err).
+				Str("op", "route").
+				Str("addr", target).
+				Msg("couldn't serialize envelope")
 			errReset := stream.Reset()
 			ignore(errReset)
-		} else {
-			routeCountSuccess.Inc()
-			stream.Close()
+			routeCount.Dec()
+			return err
+		}
+		aeaEnvelope := &dhtnode.AeaEnvelope{
+			Envel:  envelBytes,
+			Record: envelRec,
```

Comment:
[STP3.a] when a peer receives outgoing envelope it proceed to continuously lookup the dht for the destination until it